NDTensors/ext/NDTensorsMetalExt/mul.jl

```diff
@@ -1,15 +1,11 @@
 # This was calling generic matrix multiplication.
 # TODO: Raise an issue with `Metal.jl`.
-function NDTensors.mul!!(
-  ::Type{<:MtlArray},
-  CM::Transpose,
-  ::Type{<:MtlArray},
-  AM::AbstractMatrix,
-  ::Type{<:MtlArray},
-  BM::AbstractMatrix,
+function LinearAlgebra.mul!(
+  CM::Exposed{<:MtlArray,<:Transpose},
+  AM::Exposed{<:MtlArray},
+  BM::Exposed{<:MtlArray},
   α,
   β,
 )
-  mul!(transpose(CM), transpose(BM), transpose(AM), α, β)
-  return CM
+  return mul!(transpose(CM.object), transpose(BM.object), transpose(AM.object), α, β)
```

Comment:
We should explicitly use `return CM` as before.

Again, for mutating/in-place functions (`f!(dest, ...)`) we should call it in a separate line and then return the modified value `dest`, since we can't assume the `f!` will in general return `dest`. For maybe-mutating functions (`f!!(dest, ...)`)  the convention is that they need to return `dest` so then what you wrote here would be ok.

Reply:
See my comment below about conventions of functions that accept `Exposed` objects, as well as other suggestions on semantics of `Exposed` objects.

Reply:
Not sure why `mul!!` would be removed, that function is still useful for performing maybe-mutating matrix multiplication. So I don't really understand your comment but maybe my comment below will clarify some things.

Reply:
Your comment does make sense. I removed the `mul!!(CM, AM, BM, a, b)` function and replaced it with `Base.mul!(CM::Exposed, AM::Exposed, BM::Exposed, a, b)` and dispatch the correct `mul!` using the `Exposed` logic.  `mul!!` is not part of `LinearAlgebra`
```
julia> LinearAlgebra.mul!!
ERROR: UndefVarError: `mul!!` not defined
Stacktrace:
 [1] getproperty(x::Module, f::Symbol)
   @ Base ./Base.jl:31
 [2] top-level scope
   @ REPL[35]:1
   ```
but since it is  useful for the case you discuss I will add the function back into the `abstractarray` folder and define it as 
   ```
 function mul!!(CM::AbstractArray, AM::AbstractArray, BM::AbstractArray, a, b)
    mul!(expose(CM), expose(AM), expose(BM), a, b)
    return unexpose(CM)
```
Let me know if this is okay or if I should do more! Thanks for your comments!

Reply:
Sure, it's fine to have our own versions of functions that don't exist in `Base`, the goal is just to overload `Base` functions when we can.

I could imagine as an alternative that we define:
```julia
function mul!!(CM::AbstractArray, AM::AbstractArray, BM::AbstractArray, a, b)
 return mul!!(expose(CM), expose(AM), expose(BM), a, b)
end
```
and then have a version of `mul!!` defined on `Exposed` which likely then calls `mul!`. Not sure what is best here or if there is an application for that, I think it depends on how we end up using `mul!!`. As far as I know the main application we have for `mul!!` is handling the logic for allocating `UnallocatedZeros` so not sure how `Exposed` plays a role there.

Reply:
Maybe a strategy is to try to keep things similar to how we currently do things with `leaf_parenttype` dispatch.

Reply:
Okay I understand! My thinking has been to make the `Exposed` system akin to the `leaf_parenttype` system and I'm just trying to be forward thinking about the extent of the system. In my work so far I am making it as light weight as possible. Thanks again for the direction.

Right now, I am not sure that we need the `mul!!(Exposed)` version right now. With the system I currently have implemented (I will push the changes in a minute) we should be able to define something like 
```
function mul!!(CM::UnallocatedZero, AM::AbstractArray, BM::AbstractArray, a, b)
  CM = allocate(CM)
  CM = mul!(expose(CM), expose(AM), expose(BM), a, b)
  return CM
```
which can cover the `UnallocatedZeros` logic in this case.